util/filter_bench.cc

```diff
@@ -0,0 +1,403 @@
+                    fastrange32(infos.size() - numPrimaryFilters, random());
+    }
+    FilterInfo &info = infos[filterIndex];
+    if (FLAGS_new_reader_every > 0 &&
+        fastrange32(FLAGS_new_reader_every, random()) < batchSize) {
+      info.reader_.reset(policy->GetFilterBitsReader(info.filter_));
+    }
+    for (size_t i = 0; i < batchSize; ++i) {
+      if (inside) {
+        kms[i].Get(info.filter_id_, fastrange32(info.keys_added_, random()));
+      } else {
+        kms[i].Get(info.filter_id_, random() | 0x80000000);
+        info.outside_queries_++;
+      }
+    }
+    if (mode == TM_BatchPrepared && !dry_run) {
+      for (size_t i = 0; i < batchSize; ++i) {
+        batchResults[i] = false;
+      }
+      info.reader_->MayMatch(batchSize, batchSlices.get(), batchResults.get());
+      for (size_t i = 0; i < batchSize; ++i) {
+        if (inside) {
+          always_assert(batchResults[i]);
+        } else {
+          info.false_positives_ += batchResults[i];
+        }
+      }
+    } else {
+      for (size_t i = 0; i < batchSize; ++i) {
+        if (dry_run) {
+          dryRunHash ^= rocksdb::BloomHash(kms[i].slice_);
+        } else if (inside) {
+          always_assert(info.reader_->MayMatch(kms[i].slice_));
+        } else {
+          info.false_positives_ += info.reader_->MayMatch(kms[i].slice_);
+        }
+      }
+    }
+  }
+
+  uint64_t elapsedNanos = timer.ElapsedNanos();
+  double ns = (double)elapsedNanos / maxQueries;
+
+  if (dry_run) {
+    // Printing part of hash prevents dry run components from being optimized
+    // away by compiler
+    std::cout << "  Dry run (" << std::hex << (dryRunHash & 0xfff) << std::dec
+              << ") ";
+  } else {
+    std::cout << "  " << TestModeToString(mode) << " ";
+  }
+  std::cout << "ns/op: " << ns << std::endl;
+
+  if (!inside && !dry_run && mode == TM_RandomFilter) {
+    uint64_t q = 0;
+    uint64_t fp = 0;
+    double worst_fp_rate = 0.0;
+    double best_fp_rate = 1.0;
+    for (auto &info : infos) {
+      q += info.outside_queries_;
+      fp += info.false_positives_;
+      if (info.outside_queries_ > 0) {
+        double fp_rate = (double)info.false_positives_ / info.outside_queries_;
+        worst_fp_rate = std::max(worst_fp_rate, fp_rate);
+        best_fp_rate = std::min(best_fp_rate, fp_rate);
+      }
+    }
+    std::cout << "    Average FP rate %: " << 100.0 * fp / q << std::endl;
+    if (!FLAGS_quick) {
+      std::cout << "    Worst   FP rate %: " << 100.0 * worst_fp_rate
+                << std::endl;
+      std::cout << "    Best    FP rate %: " << 100.0 * best_fp_rate
+                << std::endl;
+      std::cout << "    Best possible bits/key: "
+                << -std::log((double)fp / q) / std::log(2.0) << std::endl;
+    }
+  }
+}
+
+int main(int argc, char **argv) {
+  rocksdb::port::InstallStackTraceHandler();
+  SetUsageMessage(std::string("\nUSAGE:\n") + std::string(argv[0]) +
+                  " [OPTIONS]...");
```

Comment:
It might be a good idea to mention more usage here. How to use the benchmark tool is not very intuitive to me.

Anyway, we need to document somewhere the main usage.

Reply:
You can't currently pick between different implementations, so you literally just run it. I use -quick most often (I'll add that to USAGE) but no other options seem critical to mention. Thanks to gflags, -help tells you about all the options. I'll add some info about interpreting results.

Reply:
And I think it's too niche for going on the wiki, for now. Perhaps it would make sense when/if we expose new user options to pick between SST filter implementations.